src/store.c

```diff
@@ -339,144 +374,255 @@ static void writer_close(
-        rill_ts_t ts, size_t quant,
+        rill_ts_t ts,
+        size_t quant,
         struct rill_pairs *pairs)
 {
     rill_pairs_compact(pairs);
     if (!pairs->len) return true;
 
     struct vals *vals = vals_from_pairs(pairs);
     if (!vals) goto fail_vals;
+    struct vals *invert_vals = keys_from_pairs(pairs); // TODO fix fail cleanup
 
     struct rill_store store = {0};
-    if (!writer_open(&store, file, vals, pairs->len, ts, quant)) {
+    if (!writer_open(&store, file, vals, invert_vals,
+                     pairs->len, ts, quant)) {
         rill_fail("unable to create '%s'", file);
         goto fail_open;
     }
 
-    struct indexer *indexer = indexer_alloc(pairs->len);
-    struct encoder coder = writer_begin(&store, vals, indexer);
+    struct indexer *indexer_a = indexer_alloc(pairs->len); // TODO catch fail alloc here
+    if (!indexer_a) goto fail_indexer_a_alloc;
+    struct indexer *indexer_b = indexer_alloc(pairs->len); // TODO catch fail alloc here
+    if (!indexer_b) goto fail_indexer_b_alloc;
+
+    init_store_offsets(&store, pairs->len);
+
+    struct encoder coder_a =
+        store_encoder(&store, indexer_a, vals, store.head->data_a_off);
+
+    for (size_t i = 0; i < pairs->len; ++i) {
+        if (!coder_encode(&coder_a, &pairs->data[i])) goto fail_encode_a;
+    }
+    if (!coder_finish(&coder_a)) goto fail_encode_a;
+
+    prepare_col_b_offsets(&store, &coder_a);
+
+    struct encoder coder_b =
+        store_encoder(&store, indexer_b, invert_vals, store.head->data_b_off);
+
+    rill_pairs_invert(pairs);
+    rill_pairs_compact(pairs); /* recompact mainly for sort */
 
     for (size_t i = 0; i < pairs->len; ++i) {
-        if (!coder_encode(&coder, &pairs->data[i])) goto fail_encode;
+        if (!coder_encode(&coder_b, &pairs->data[i])) goto fail_encode_b;
     }
-    if (!coder_finish(&coder)) goto fail_encode;
+    if (!coder_finish(&coder_b)) goto fail_encode_b;
+
+    writer_flush_indices(&store, indexer_a, indexer_b);
+
+    store.head->pairs = coder_a.pairs;
 
-    store.head->keys = coder.keys;
-    store.head->pairs = coder.pairs;
+    writer_close(&store, store.head->data_b_off + coder_off(&coder_b));
 
-    writer_close(&store, indexer, store.head->data_off + coder_off(&coder));
+    coder_close(&coder_a);
+    coder_close(&coder_b);
+
+    indexer_free(indexer_a);
+    indexer_free(indexer_b);
 
-    coder_close(&coder);
-    indexer_free(indexer);
     free(vals);
+    free(invert_vals);
+
     return true;
 
-  fail_encode:
-    coder_close(&coder);
-    writer_close(&store, indexer, 0);
-    indexer_free(indexer);
+  fail_encode_b:
+    coder_close(&coder_b);
+    writer_close(&store, 0);
+  fail_encode_a:
+    coder_close(&coder_a);
+    indexer_free(indexer_b);
+  fail_indexer_b_alloc:
+    indexer_free(indexer_a);
+  fail_indexer_a_alloc:
   fail_open:
     free(vals);
   fail_vals:
     return false;
 }
 
+static struct vals *vals_from_index(struct index* index)
```

Comment:
This function sounds like it should be in either `vals.c` or `index.c`

Reply:
I've decided not to move this function. If I move it to, let's say index, which is the logical place for that to reside, I need to expose the function in `rill.h`, and, then further expose the `index struct`. I would be against bleeding too much out of the interface, unless you really want this out of `store.c`.

Reply:
Update - this function is no longer needed due to implementation of your suggestion here: https://github.com/RAttab/rill/pull/6#discussion_r177229898 (fixorzed? fixorzed.)